src/rgw/rgw_sal_motr.cc

```diff
@@ -67,6 +68,89 @@ static unsigned roundup(unsigned x, unsigned by)
   return ((x - 1) / by + 1) * by;
 }
 
+std::string base62_encode(uint64_t value, size_t pad)
+{
+  // Integer to Base62 encoding table. Characters are sorted in
+  // lexicographical order, which makes the encoded result
+  // also sortable in the same way as the integer source.
+  constexpr std::array<char, 62> base62_chars{
+      // 0-9
+      '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+      // A-Z
+      'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+      'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+      // a-z
+      'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+      'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'};
+
+  std::string ret;
+
+  if (value == 0) {
+    ret = base62_chars[0];
+  }
+
+  while (value > 0) {
+    ret = base62_chars[value % base62_chars.size()] + ret;
```

Comment:
**Just a Suggestion**
prepending char each time is like creating a temp string each time and assigning it to ret again.
Instead, we can append the char and reverse the base62_chars array, which will produce the same results.
```c++
constexpr std::array<char, 62> base62_chars{
    // z-a
    // Z-A
    // 9-0
};
...
while (value > 0) {
    ret += base62_chars[value % base62_chars.size()];
    value /= base62_chars.size();
}
```

Reply:
good catch @sumedhak27. However i feel it will still result in copying into new string, on every append.
Instead, since we know finite length of version-id here, we can use that to statically assign memory for required str length

Reply:
If we know the exact size then we can reserve the space directly that will be great!! 😃 

`However i feel it will still result in copying into new string, on every append.`
true... just that size gets incremented in powers of 2, just like vectors.

Reply:
> true... just that size gets incremented in powers of 2, just like vectors.

true, but its also said prior memory allocation differs for different implementations.
check https://www.oreilly.com/library/view/optimized-c/9781491922057/ch04.html (Strings Are Dynamically Allocated)

Reply:
> https://www.oreilly.com/library/view/optimized-c/9781491922057/ch04.html (Strings Are Dynamically Allocated)

Thanks, @cdeshmukh , this was a great read!! It also justified & answered some of my doubts. 😄

> Simply reversing the string is not the same. Consider a simple integer value, 0. In the original encoding, it becomes "0". In your proposal it becomes "z"

Correct, sorry, I should have tried an example. 
Then the only optimization we can easily do is probably reserve some length for string, append chars and then reverse the string, but that too as you said might not show considerable improvement in practice. Thanks, @keithpine for the explanation, I don't really have any knowledge about base62_encoding, I just started learning about it, I had read about string optimizations during my studies, so was trying to apply it. 😅

Reply:
I think it would be pretty easy to benchmark a couple of implementations with a simple test program (e.g. unit test) and compare the results. Just do several runs, generating N number of version IDs each time. If there's any significant difference between them, on average, pick the best one.

Thinking about optimizations is perfectly fine, but it's not always obvious what the compiler ends up doing. The article Chetan linked to is good, and seems to be pretty recent. The different approaches in that article are also backed by benchmarks.

Inserting to the front of a string _seems_ like an obvious bad idea in hindsight, but for version IDs we're only dealing with 8 characters at most, so it wasn't a big concern of mine at the time. Feel free to change the implementation any way you like, I'm not attached to it. 👍

Reply:
@shriya-deshmukh is there any typo above? how second approach time 0.058726ms is better than first approach time 0.045644ms ?

Reply:
Oh yes sorry, my bad, first approach is taking more time than the 2nd approach.

1.existing approach - prepending char to the string (execution time 0.089015ms)
2. append char to string and then reverse the string (execution time 0.045699ms)